crates/vm/levm/src/call_frame.rs

```diff
@@ -164,6 +152,36 @@ impl Stack {
     pub fn clear(&mut self) {
         self.offset = STACK_LIMIT;
     }
+
+    /// Pushes a copy of the value at depth N
+    #[inline]
+    pub fn dup<const N: usize>(&mut self) -> Result<(), ExceptionalHalt> {
+        // Compile-time check that ensures `self.offset + N` is safe,
+        // since self.offset is bounded by STACK_LIMIT
+        const {
+            assert!(STACK_LIMIT.checked_add(N).is_some());
+        }
+        #[expect(clippy::arithmetic_side_effects)]
+        let index = self.offset + N;
+        if index >= self.values.len() {
+            return Err(ExceptionalHalt::StackUnderflow);
+        }
+
+        self.offset = self
+            .offset
+            .checked_sub(1)
+            .ok_or(ExceptionalHalt::StackOverflow)?;
+
+        #[expect(unsafe_code, reason = "index < size, offset-1 >= 0")]
+        unsafe {
+            std::ptr::copy_nonoverlapping(
+                self.values.get_unchecked_mut(index).0.as_mut_ptr(),
+                self.values.get_unchecked_mut(self.offset).0.as_mut_ptr(),
+                4,
```

Comment:
Can we use a constant for this? Otherwise, let's put it in a local variable or add a comment.

Reply:
Right. I mixed it up with `op_swap`, which does start at 1.